x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskRandomTests.java

```diff
@@ -0,0 +1,282 @@
+                        }
+                        indexMetadataVersion = testResponse.indexMetadataVersion;
+                        if (testResponse.exception != null) {
+                            errorHandler.accept(testResponse.exception);
+                        } else {
+                            handler.accept(testResponse.response);
+                        }
+                    } else {
+                        assert from >= testRun.finalExpectedGlobalCheckpoint;
+                        handler.accept(new ShardChangesAction.Response(0L, tracker.getCheckpoint(), new Translog.Operation[0]));
+                    }
+                };
+                threadPool.generic().execute(task);
+            }
+
+            @Override
+            protected boolean isStopped() {
+                return stopped.get();
+            }
+
+            @Override
+            public void markAsCompleted() {
+                stopped.set(true);
+                tearDown();
+            }
+
+            @Override
+            public void markAsFailed(Exception e) {
+                stopped.set(true);
+                tearDown();
+            }
+
+            private void tearDown() {
+                threadPool.shutdown();
+                List<Translog.Operation> expectedOperations = testRun.responses.values().stream()
+                    .flatMap(List::stream)
+                    .map(testResponse -> testResponse.response)
+                    .filter(Objects::nonNull)
+                    .flatMap(response -> Arrays.stream(response.getOperations()))
+                    .sorted(Comparator.comparingLong(Translog.Operation::seqNo))
+                    .collect(Collectors.toList());
+                assertThat(receivedOperations.size(), equalTo(expectedOperations.size()));
+                receivedOperations.sort(Comparator.comparingLong(Translog.Operation::seqNo));
+                for (int i = 0; i < receivedOperations.size(); i++) {
+                    Translog.Operation actual = receivedOperations.get(i);
+                    Translog.Operation expected = expectedOperations.get(i);
+                    assertThat(actual, equalTo(expected));
+                }
+            }
+        };
+    }
+
+    private static TestRun createTestRun(long startSeqNo, long startIndexMetadataVersion, int maxOperationCount) {
+        long prevGlobalCheckpoint = startSeqNo;
+        long indexMetaDataVersion = startIndexMetadataVersion;
+        int numResponses = randomIntBetween(16, 256);
+        Map<Long, List<TestResponse>> responses = new HashMap<>(numResponses);
+        for (int i = 0; i < numResponses; i++) {
+            long nextGlobalCheckPoint = prevGlobalCheckpoint + maxOperationCount;
+            if (sometimes()) {
+                indexMetaDataVersion++;
+            }
+
+            if (sometimes()) {
+                List<Translog.Operation> ops = new ArrayList<>();
```

Comment:
Can we move this `ops` just before we generate the actual response (after the error response). It may be easier to follow.